Vulkan AS rebuild-on-replay: BDA tracking

cmannett85-arm opened the pull request:
In order to unify the input structs between host and device builds, the AS API uses addresses rather than VkBuffer handles, but there's no way of retrieving a handle from an address so we have to track all VkBuffer binds to map their addresses back to the owning handle. This is done is in TrackInputBuffer(..) and UntrackInputBuffer(..), the latter for when the resource is destroyed. A complexity here is that when doing the lookup, users are allowed to manipulate the BDA so we have to find the buffer the address lies in and return the offset too.

baldurk commented on renderdoc/driver/vulkan/vk_acceleration_structure.h:
```diff
@@ -37,7 +43,18 @@ class VulkanAccelerationStructureManager
+  // ResourceId.  BDA isn't known at creation time and the usage flags aren't known at bind time so
+  // initialise is a two-step process.  These should only be called when capturing
+  void TrackInputBuffer(VkDevice device, VkResourceRecord *record);
+  void UntrackInputBuffer(VkResourceRecord *record);
```
Rather than having a new implementation of this, I think it would be better to move D3D12's address tracking to a new common location and then use it here.

If you look at `GPUAddressRangeTracker` in `d3d12_manager.h` it should be fairly easy to lift out - `D3D12_GPU_VIRTUAL_ADDRESS` is a typedef for `uint64_t` the same as `VkDeviceAddress` so we can define an API-agnostic typedef and use it with implicit conversions. We don't have much that is shared between APIs but it can be moved to a new pair of files next to `intervals.h/cpp`.

Also the likelihood is high that in the future we'll need to do this kind of lookup if the API design continues its trend - can you put this tracking inside `WrappedVulkan` itself rather than tying it to acceleration structures as this in itself is not AS-specific.

baldurk commented on renderdoc/driver/vulkan/vk_acceleration_structure.h:
```diff
@@ -37,7 +43,18 @@ class VulkanAccelerationStructureManager
+  // are copied.
+  void CopyInputBuffers(VkCommandBuffer commandBuffer,
+                        const VkAccelerationStructureBuildGeometryInfoKHR &info,
+                        const VkAccelerationStructureBuildRangeInfoKHR *buildRange);
```
This function is unused/orphaned in this PR so best to leave it out for now.

cmannett85-arm replied:
It is used but only to provide the built flag in the AS' `VkResourceRecord`, which is then set true upon queue submission.  Yes it's a bit arbitrary now but it lays down the architecture for the subsequent patches.

baldurk replied:
Oh I see, this should be done differently - the cold data (`accelerationStructureInfo` in this case) should be initialised at creation time, so in `WrappedVulkan::vkCreateAccelerationStructureKHR` where the resource record is created. That would mean this function should still be deferred to a later patch.

baldurk commented on renderdoc/driver/vulkan/wrappers/vk_resource_funcs.cpp:
```diff
@@ -1525,6 +1525,9 @@ VkResult WrappedVulkan::vkBindBufferMemory(VkDevice device, VkBuffer buffer, VkD
     // appropriate.
     GetResourceManager()->MarkDirtyResource(GetResID(memory));
+
+    if(m_DeviceAddressResources.IDs.contains(record->GetResourceID()))
```
This access needs to lock around the lookup or else it could be racing against an ID being inserted during creation/allocation on another thread.

cmannett85-arm commented on renderdoc/core/gpu_address_range_tracker.cpp:
```diff
@@ -0,0 +1,137 @@
+
+  GPUAddressRange range;
+
+  // this should really be a read-write lock
```
@baldurk is this comment still valid?  My reading of the code is that read lock is sufficient as nothing is being written back to the array.

baldurk replied:
Indeed this comment looks to be stale and can be removed - I think it was meaning the lock itself should be a read-write lock.

baldurk commented on renderdoc/api/replay/resourceid.h:
```diff
@@ -24,6 +24,8 @@
 
 #pragma once
 
+#include <functional>
```
Please don't add a large STL include like this to what is (practically) a global header. If this is for the `lower_bound` calls in gpu_address_range_tracker.cpp then it should be included only in that cpp.

cmannett85-arm replied:
It's not for that, it's because `ResourceId` has a `friend struct std::hash<ResourceId>` declaration in it, but `std::hash` hasn't been declared at this point.  It would be possible to forward declare everything and then implement the `std::hash<ResourceId>()` operator in a source file after including one of the appropriate std headers - there would be a performance hit for this though.

A medium solution would be to move the include to above the `ResourceId` include in `gpu_address_range_tracker.h`, it's still commonly included (as `vk_core.h` and `d3d12_manager.h` include it) though.

baldurk replied:
Oh I see, then that's my mistake - I thought I'd expunged most of these heavyweight STL headers but some are still needed. I guess this wasn't present before because usually code doesn't include just `resourceid.h` on its own. This is fine then.

baldurk commented on renderdoc/driver/vulkan/vk_memory.cpp:
```diff
@@ -28,6 +28,66 @@
 
+void WrappedVulkan::TrackBufferAddress(VkDevice device, VkBuffer buffer)
+{
+  SCOPED_LOCK(m_DeviceAddressResourcesLock);
```
As a minor improvement it would be better to reduce contention on this lock by just doing the check while holding the lock rather than all of the rest of the work, e.g:

```cpp
bool isBDA = false;
{
  SCOPED_LOCK(m_DeviceAddressResourcesLock);
  isBDA = m_DeviceAddressResources.IDs.contains(GetResID(buffer));
}
```

cmannett85-arm committed:
Vulkan AS rebuild-on-replay: BDA tracking

In order to unify the input structs between host and device builds, the AS API uses addresses rather than VkBuffer handles, but there's no way of retrieving a handle from an address so we have to track all VkBuffer binds to map their addresses back to the owning handle. This is done is in TrackInputBuffer(..) and UntrackInputBuffer(..), the latter for when the resource is destroyed. A complexity here is that when doing the lookup, users are allowed to manipulate the BDA so we have to find the buffer the address lies in and return the offset too.

cmannett85-arm commented:
@baldurk is this sufficient, or would you like any other changes?
